tests/serialization/test_dag_serialization.py

```diff
@@ -292,6 +292,55 @@ def validate_operator_extra_links(self, task):
             "https://www.google.com"
         )
 
+    @parameterized.expand([
+        (datetime(2019, 8, 1), None, datetime(2019, 8, 1)),
+        (datetime(2019, 8, 1), datetime(2019, 8, 2), datetime(2019, 8, 2)),
+        (datetime(2019, 8, 1), datetime(2019, 7, 30), datetime(2019, 8, 1)),
```

Comment:
I'm not sure about this test case - shouldn't the start_date attributes in the Task be whatever was set?

Reply:
dag.add_task changes the start_date as:

task.start_date=max(task.start_date, dag.start_date)

Reply:
Oh, right.

I wonder if it's worth calling `dag.add_tasks` in our deserializer instead of what we are currently doing?

(As the other thing that we don't set currently is `dag.task_count`)

Reply:
Yes, we can do that too. The following should work in theory:

```diff
diff --git a/airflow/serialization/serialized_dag.py b/airflow/serialization/serialized_dag.py
index 1f6d3b331..61b75e41a 100644
--- a/airflow/serialization/serialized_dag.py
+++ b/airflow/serialization/serialized_dag.py
@@ -100,10 +100,8 @@ class SerializedDAG(DAG, Serialization):
             if k == "_downstream_task_ids":
                 v = set(v)
             elif k == "tasks":
-                v = {
-                    task["task_id"]: SerializedBaseOperator.deserialize_operator(task) for task in v
-                }
-                k = "task_dict"
+                dag.add_tasks([SerializedBaseOperator.deserialize_operator(task) for task in v])
+                continue
             elif k == "timezone":
                 v = cls._deserialize_timezone(v)
             elif k in {"retry_delay", "execution_timeout"}:
@@ -124,13 +122,8 @@ class SerializedDAG(DAG, Serialization):

         setattr(dag, 'full_filepath', dag.fileloc)
         for task in dag.task_dict.values():
-            task.dag = dag
             task = cast(SerializedBaseOperator, task)

-            for date_attr in ["start_date", "end_date"]:
-                if getattr(task, date_attr) is None:
-                    setattr(task, date_attr, getattr(dag, date_attr))
-
             if task.subdag is not None:
                 setattr(task.subdag, 'parent_dag', dag)
                 task.subdag.is_subdag = True
```

I have no idea why do we set `dag.task_count` as it is not used anywhere !

Reply:
We can actually also do the following since we are already looping across all tasks:

```diff
diff --git a/airflow/serialization/serialized_dag.py b/airflow/serialization/serialized_dag.py
index 1f6d3b331..e0f761ddd 100644
--- a/airflow/serialization/serialized_dag.py
+++ b/airflow/serialization/serialized_dag.py
@@ -100,10 +100,9 @@ class SerializedDAG(DAG, Serialization):
             if k == "_downstream_task_ids":
                 v = set(v)
             elif k == "tasks":
-                v = {
-                    task["task_id"]: SerializedBaseOperator.deserialize_operator(task) for task in v
-                }
-                k = "task_dict"
+                for task in v:
+                    dag.add_task(SerializedBaseOperator.deserialize_operator(task))
+                continue
             elif k == "timezone":
                 v = cls._deserialize_timezone(v)
             elif k in {"retry_delay", "execution_timeout"}:
@@ -124,13 +123,8 @@ class SerializedDAG(DAG, Serialization):

         setattr(dag, 'full_filepath', dag.fileloc)
         for task in dag.task_dict.values():
-            task.dag = dag
             task = cast(SerializedBaseOperator, task)

-            for date_attr in ["start_date", "end_date"]:
-                if getattr(task, date_attr) is None:
-                    setattr(task, date_attr, getattr(dag, date_attr))
-
             if task.subdag is not None:
                 setattr(task.subdag, 'parent_dag', dag)
                 task.subdag.is_subdag = True
```